psi4/src/core.cc

```diff
@@ -829,33 +826,19 @@ bool py_psi_set_global_option_array(std::string const& key, py::list values, Dat
         if (data.type() == "array")
             Process::environment.options.set_global_array(nonconst_key);
     }
-    size_t size = len(values);
-    for (int n = 0; n < size; ++n) {
-        try {
-            py::list l = values[n].cast<py::list>();
-            DataType *newentry = Process::environment.options.set_global_array_array(nonconst_key, entry);
-            // Now we need to recurse, to fill in the data
-            py_psi_set_global_option_array(key, l, newentry);
-        }
-        catch (py::cast_error e) {
-            // This is not a list; try to cast to a string
-            try {
-                std::string s = values[n].cast<std::string>();
-                Process::environment.options.set_global_array_string(nonconst_key, s, entry);
-            } catch(py::cast_error e) {
-                try {
-                    // This is not a list or string; try to cast to an integer
-                    int i = values[n].cast<int>();
-                    Process::environment.options.set_global_array_int(nonconst_key, i, entry);
-                } catch(py::cast_error e) {
-                    // This had better be castable to a float.  We don't catch the exception here
-                    // because if we encounter one, something bad has happened
-                    double f = values[n].cast<double>();
-                    Process::environment.options.set_global_array_double(nonconst_key, f, entry);
-                }
-            }
-
+    Options& options=Process::environment.options;
+    for(py::handle oi: values){
+        if(can_cast<py::list>(oi)){
+            DataType* newentry = options.set_global_array_array(nonconst_key, entry);
+            py_psi_set_global_option_array(key,oi.cast<py::list>(),newentry);
         }
+        else if(can_cast<std::string>(oi))
+            options.set_global_array_string(nonconst_key, oi.cast<std::string>(), entry);
+        else if(can_cast<int>(oi))
+            options.set_global_array_int(nonconst_key, oi.cast<int>(), entry);
+        else if(can_cast<double>(oi))
+            options.set_global_array_double(nonconst_key, oi.cast<double>(), entry);
+        else throw PSIEXCEPTION("unrecognized type in py_psi_set_global_option_array");
```

Comment:
You checked all this on the [options test case](https://github.com/psi4/psi4/blob/master/tests/options1/input.dat)?